contracts/extensions/ede007-snapshot-proposal-voting-v4.clar

```diff
@@ -0,0 +1,150 @@
+;; Custom majority thresholds for voting are also possible on a per proposal basis.
+;; A custom majority of 66% mean the percent of votes for must be greater than 66 for
+;; the vote to carry.
+
+(impl-trait .extension-trait.extension-trait)
+(use-trait proposal-trait .proposal-trait.proposal-trait)
+
+(define-constant err-unauthorised (err u3000))
+(define-constant err-proposal-already-executed (err u3001))
+(define-constant err-proposal-already-exists (err u3002))
+(define-constant err-unknown-proposal (err u3003))
+(define-constant err-proposal-already-concluded (err u3004))
+(define-constant err-proposal-inactive (err u3005))
+(define-constant err-insufficient-voting-capacity (err u3006))
+(define-constant err-end-block-height-not-reached (err u3007))
+(define-constant err-not-majority (err u3008))
+(define-constant err-exceeds-voting-cap (err u3009))
+
+(define-constant custom-majority-upper u10000)
+;; (define-constant vote-cap u2083333333333)
+
+(define-map proposals
+	principal
+	{
+		votes-for: uint,
+		votes-against: uint,
+		start-block-height: uint,
+		end-block-height: uint,
+		concluded: bool,
+		passed: bool,
+		custom-majority: (optional uint), ;; u10000 = 100%
+		proposer: principal
+	}
+)
+
+(define-map member-total-votes {proposal: principal, voter: principal} uint)
+
+;; --- Authorisation check
+
+(define-public (is-dao-or-extension)
+	(ok (asserts! (or (is-eq tx-sender .ecosystem-dao) (contract-call? .ecosystem-dao is-extension contract-caller)) err-unauthorised))
+)
+
+;; --- Internal DAO functions
+
+;; Proposals
+
+(define-public (add-proposal (proposal <proposal-trait>) (data {start-block-height: uint, end-block-height: uint, proposer: principal, custom-majority: (optional uint)}))
+	(begin
+		(try! (is-dao-or-extension))
+		(asserts! (is-none (contract-call? .ecosystem-dao executed-at proposal)) err-proposal-already-executed)
+		(asserts! (match (get custom-majority data) majority (> majority u5000) true) err-not-majority)
+		(print {event: "propose", proposal: proposal, proposer: tx-sender})
+		(ok (asserts! (map-insert proposals (contract-of proposal) (merge {votes-for: u0, votes-against: u0, concluded: false, passed: false} data)) err-proposal-already-exists))
+	)
+)
+
+;; --- Public functions
+
+;; Proposals
+
+(define-read-only (get-proposal-data (proposal principal))
+	(map-get? proposals proposal)
+)
+
+;; Votes
+
+(define-read-only (get-current-total-votes (proposal principal) (voter principal))
+	(default-to u0 (map-get? member-total-votes {proposal: proposal, voter: voter}))
+)
+
+(define-read-only (get-historical-values (height uint) (who principal))
+	(at-block (unwrap! (get-block-info? id-header-hash height) none)
+		(some {
+				user-balance: (stx-get-balance who), 
+				locked-balance: (default-to u0 (get amount-ustx (contract-call? 'SP000000000000000000002Q6VF78.pox get-stacker-info who))), 
+				voting-cap: (contract-call? 'SP000000000000000000002Q6VF78.pox get-stacking-minimum)
```

Comment:
voting-cap is constant per proposal but varies between proposals. It could be stored per by `submit-proposal-for-vote` proposal but this would bloat contract storage as its only needed runtime.